Microsoft.HealthVault.Fhir/Transformers/ObservationToHealthVault.cs

```diff
@@ -68,22 +71,37 @@ public static ThingBase ToHealthVault(this Observation observation)
 
         internal static T GetThingValueFromQuantity<T>(Quantity value) where T : Measurement<double>, new()
         {
-            // TODO: detect the units from the code (value.Unit)
-            if (value != null)
+            if (value?.Value == null)
             {
-                if (value.Value.HasValue)
-                {
-                    var result = new T();
-                    result.Value = (double)value.Value;
-                    result.DisplayValue = new DisplayValue();
-                    result.DisplayValue.Value = (double)value.Value;
-                    result.DisplayValue.Units = value.Unit;
-                    result.DisplayValue.UnitsCode = value.Code;
-                    return result;
-                }
+                return null;
             }
+            
+            var unitConversion = UnitResolver.Instance.UnitConversions.FirstOrDefault(x => x.Code == value.Code);
 
-            return null;
+            double convertedValue;
+            if (unitConversion != null)
+            {
+                var unitEnum = Enum.Parse(Type.GetType($"UnitsNet.Units.{unitConversion.UnitsNetUnitEnum}, UnitsNet"), unitConversion.UnitsNetSource);
```

Comment:
Agreed, TryParse should be used here

---
In reply to: [127042699](https://github.com/Microsoft/healthvault-fhir-library/pull/7#discussion_r127042699) [](ancestors = 127042699)